p2p/transport/websocket/websocket_test.go

```diff
@@ -549,3 +552,151 @@ func TestResolveMultiaddr(t *testing.T) {
 		})
 	}
 }
+
+func TestReusePortOnDial(t *testing.T) {
+
+	// Create an endpoint that will accept connections.
+	// We'll use this to verify that the party initiating the connection reused port.
+	serverID, cu := newUpgrader(t)
+	transport, err := New(cu, &network.NullResourceManager{})
+	require.NoError(t, err)
+
+	server, err := transport.Listen(ma.StringCast("/ip4/127.0.0.1/tcp/0/ws"))
+	require.NoError(t, err)
+	defer server.Close()
+
+	// Create an endpoint that will initiate connection.
+	_, u := newUpgrader(t)
+	tpt, err := New(u, &network.NullResourceManager{}, EnableReuseport())
+	require.NoError(t, err)
+
+	// Start listening.
+	listener, err := tpt.Listen(ma.StringCast("/ip4/127.0.0.1/tcp/0/ws"))
+	require.NoError(t, err)
+	defer listener.Close()
+
+	// Take a note of the multiaddress on which we listen. This should be the address from which we dial too.
+	expectedAddr := listener.Multiaddr()
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		conn, err := server.Accept()
+		require.NoError(t, err)
+		defer conn.Close()
+
+		// The meat of this test - verify that the connection was received from the same port as the listen port recorded above.
+		remote := conn.RemoteMultiaddr()
+		require.Equal(t, expectedAddr, remote)
+	}()
+
+	conn, err := tpt.Dial(context.Background(), server.Multiaddr(), serverID)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	<-done
+}
+
+func TestReusePortOnListen(t *testing.T) {
+
+	const (
+		// how many connections we try to establish.
+		connectionCount = 20
+	)
+
+	// Create an endpoint that will accept connections.
+	// We'll use this to verify that the party initiating the connection reused port.
+	_, cu := newUpgrader(t)
+	tpt, err := New(cu, &network.NullResourceManager{}, EnableReuseport())
+	require.NoError(t, err)
+
+	listener1, err := tpt.maListen(ma.StringCast("/ip4/127.0.0.1/tcp/0/ws"))
+	require.NoError(t, err)
+
+	// Get the port on which we should start the second listener
+	addr, ok := listener1.Addr().(*net.TCPAddr)
+	require.True(t, ok)
+
+	port := addr.Port
+	listener2, err := tpt.maListen(ma.StringCast(fmt.Sprintf("/ip4/127.0.0.1/tcp/%v/ws", port)))
+	require.NoError(t, err)
+
+	listeners := []manet.Listener{listener1, listener2}
+
+	// Record which listener accepted how many connections.
+	requestCount := make(map[int]int)
+	var lock sync.Mutex
+
+	var connsHandled sync.WaitGroup
+	connsHandled.Add(connectionCount)
+	// For both listeners spin up goroutines to accept incoming connections.
+	for i, listener := range listeners {
+		for j := 0; j < connectionCount; j++ {
+			go func(index int, listener manet.Listener) {
+
+				conn, err := listener.Accept()
+				if err != nil {
+					// Stop condition - this happens when the listener is closed.
+					require.ErrorIs(t, err, transport.ErrListenerClosed)
+					return
+				}
```

Comment:
It's possible to just start listeners above and defer `Close()` calls to it. Then when the test function completes and execution "falls off the end" the `listener.Close()` call will unblock this and `Accept()` call will fail with an error.

Still, I think I prefer that one can see what will happen during the test just by reading it. To me it's better than "something will happen after the test is done too". This is a bit ugly but explicit. I'm open to any criticism :)